### PR TITLE
add custom CSR support to Spike

### PR DESCRIPTION
**Pull request by udinator**
This PR adds the cpuctrl and secureseed CSRs to Spike and adds two new
commandline arguments '--secure-ibex' and '--icache-en' that are needed
to properly emulate the read behavior of these CSRs.

Note that this PR does not implement the actual security-related behaviors that these CSRs control,
this is just to enable the correct read/write behavior.

Signed-off-by: Udi Jonnalagadda <udij@google.com>

### CODE REVIEW

File: riscv/processor.cc
Comment (rswarbrick) [inline riscv/processor.cc:1132]: Nit: Maybe move these to the end of the switch? It doesn't look like the existing entries are in order anyway, so maybe it makes more sense to have "Ibex-specific stuff" grouped at the bottom?

**Review by rswarbrick — APPROVED** [review]
Looks good for now, with a possible minor nit.

It would be good to "upstream" this, if possible. I suspect the right thing to do is teach the existing extension mechanism to allow adding CSRs. Then each CSR-related piece of code would do a "is there a custom CSR defined?" check before doing its thing. Does that make sense to you? If so, let's open a tracking issue and reach out to Jim Wilson, the main Spike maintainer, about it.

File: riscv/processor.cc
Comment (udinator) [reply riscv/processor.cc:1132]: Sure, can definitely move them to the end.

**Comment by udinator** [comment]
> Looks good for now, with a possible minor nit.
> 
> It would be good to "upstream" this, if possible. I suspect the right thing to do is teach the existing extension mechanism to allow adding CSRs. Then each CSR-related piece of code would do a "is there a custom CSR defined?" check before doing its thing. Does that make sense to you? If so, let's open a tracking issue and reach out to Jim Wilson, the main Spike maintainer, about it.

Thanks for the review!

Sorry, I'm not quite sure I fully understand your point - the existing extension mechanism should already allow users to easily add custom CSR definitions by 1) using the macros to declare them and 2) implementing the appropriate read/write behavior in `get_csr()` and `set_csr()` respectively. I'm not sure how "upstream-able" this sort of thing would be, since users would be explicitly declaring their platform-specific CSRs.
Could you elaborate a bit on why the CSR-related code should ask "is there a custom CSR defined" ?

**Comment by rswarbrick** [comment]
My point is that we shouldn't have a forked Spike if we can avoid it. Spike has an extension mechanism to allow you to define instruction set extensions in a library (which you load at runtime). I'm trying to work out whether we could do something similar to add custom CSRs, probably by adding to the extension interface. If that goes upstream, we can run a vanilla version of spike again, loading in the magic Ibex module.

**Comment by imphil** [comment]
I filed https://github.com/lowRISC/ibex/issues/1131 to evaluate if we can get away without patching Spike. In the mean time, taking the approach in this PR sounds sensible. @udinator can you go ahead and merge this change if you feel it is ready?

**Pull request merged**
